pubsub.go

```diff
@@ -163,6 +177,38 @@ func (p *PubsubValueStore) GetValue(ctx context.Context, key string, opts ...rop
 	return p.getLocal(key)
 }
 
+func (p *PubsubValueStore) SearchValue(ctx context.Context, key string, opts ...ropts.Option) (<-chan []byte, error) {
+	if err := p.Subscribe(key); err != nil {
+		return nil, err
+	}
+
+	p.watchLk.Lock()
+	wg, ok := p.watching[key]
+	if !ok {
+		wg = &watchGroup{
+			closing:   make(chan struct{}),
+			listeners: map[chan<- []byte]context.Context{},
+		}
+		p.watching[key] = wg
+	}
+	// Lock searchgroup before checking local storage so we don't miss updates
+	wg.lk.Lock()
+	p.watchLk.Unlock()
+
+	out := make(chan []byte, 1)
+	lv, err := p.getLocal(key)
+	if err == nil {
+		out <- lv
+	}
+
+	p.watchLk.Lock()
```

Comment:
This can deadlock.  Someone else can be holding `p.watchLk.Lock()`, waiting on `wg.lk.Lock()` (and *we're* holding `w.lk.Lock()`. Do we really need `p.watchLk`?

Reply:
This shouldn't happen anywhere (note that `Close` drops this lock before getting group lock).

`p.watchLk` guards the `watching` map, which we need to do in registration. I could re-use `p.mx`, but I'd rather have more granular locks.

Reply:
Two concurrent `SearchValue` calls for the same key should trigger this. That is, another thread could be waiting at https://github.com/libp2p/go-libp2p-pubsub-router/pull/7/files/7365dc9001d12108172504c09010ff4563f1b40f#diff-16720ecd1b41b213c82a47c8271e3096R195

(basically, this violates the "thou shalt not take locks in different orders" rule)